### DISCUSSION TITLE
How to switch an InputMask pattern while user is typing? (credit card masks)

### DISCUSSION THREAD

**Discussion by wlannistamps** [primefaces/community#2072] [PrimeReact] [answered]
I'm trying to change the value of `mask` of the InputMask while a user is typing in a credit card number. Essentially switching the pattern between Visa and Amex patterns.

My creditcardnumber component is using onChange to capture the user's input. The `mask`'s value is a prop passed in to the jsx from the component. 

On each keystroke, I run a function that does a regex test and returns a numeric value. That value gets stored as `ccType` in Redux state using a dispatch. It really only changes on the first or second keystrokes, depending on the number (Visa starts with a '4', so ccType changes on the first keystroke, whereas amex and mastercard don't trigger a change until the second keystroke, and Discover on the 4th keystroke).

I have subscribed to `ccType` using Redux's useSelector() function. Per the documentation, if that value changes, the component should re-render. So I have a useEffect() tracking ccType, and on its change, I run the function that updates the value of `mask` with the new pattern on that change.

It doesn't work. If I am updating the handleCreditCardMask() function by just checking on ccTypeState and returning a value, I get an error that it can't split the props. If I set the value of the mask to a variable in the component, and then use `useEffect()` to update the value of the mask when my ccType changes state, then pass that in as a prop, nothing happens at all, the mask never updates.

In other cases as soon as the `ccType` changes and the mask updates, the input field's value is wiped completely clean. (I have `autoclear={false}`). In the case of Visa, if I type `4` the input clears out and no mask is shown. If I type it again, the number 4 appears and the mask appears. However, if I try any of the other patterns, as soon as I hit the regex pattern that updates `ccType`, the input clears and if I keep typing it basically starts over.

How does one update the mask pattern while a user is typing?

Below is one of cases I tried

```
const CreditCardNumber = () => {

	const ccTypeState = useSelector(state => state.session.reg.ccType); // subscribes to the ccType via redux
	let mask = '9999 9999 9999 9999'; // default mask

	// When ccTypeState updates, update the credit card mask. 
	useEffect(() => {
		handleCreditCardMask(ccTypeState);
	}, [ccTypeState]);

	/**
	* Gets the credit card type based on the credit card number
	* @param {string} ccNumber - Credit card number
	* @return {number} - Credit card type
	*/
	const getCreditCardTypeByNumber = (ccNumber) => {
	    if ((/^(34|37)/).test(ccNumber)) {
		    //AMEX begins with 34 or 37
		    return 3;
	    } else if ((/^(51|52|53|54|55)/).test(ccNumber)) {
		    //MasterCard begins with 51-55
		    return 2;
	    } else if ((/^(222[1-9]|22[3-9]\\d|2[3-6]\\d{2}|27[0-1]\\d|2720)[0-9]{12}$/).test(ccNumber.replace(/\s+/g, ''))) {
		    //MasterCard begins with 2221-2720
		    return 2;
	    } else if ((/^(4)/).test(ccNumber)) {
		    //VISA begins with 4
		    return 1;
	    } else if ((/^(6011)/).test(ccNumber)) {
		    //Discover begins with 6011
		    return 4;
	    } else {
		    return null;
	    }
	}

	/**
	 * Update the credit card mask
	 * @summary Gets the credit card type updates the credit card mask
	 * @param {number} ccType - The credit card type
	 * @return {number} - The credit card type
	 */
	const handleCreditCardMask = (ccType) => {
		switch (ccType) {
			case 1: // Visa
				mask = '9999 9999 9999 9999 999';
				break;
			case 2: // Mastecard
				mask = '9999 9999 9999 9999';
				break;
			case 3: // American Express
				mask = '9999 999999 99999';
				break;
			case 4: //Discover
				mask = '9999 9999 9999 9999';
				break;
			default: // anything else
				mask = '9999 9999 9999 9999';
				break;
		}
	}

	/**
	 * Handle input change of credit card number field
	 * @summary Prevent anything but numbers (and space for masking) from being typed then trigger inputChange.
	 * Uses event.originalEvent.nativeEvent to get the actual event object, since this is a PrimeReact <InputMask/> component.
	 * @param {Event} event - The event object
	 */
	const handleInputChange = (event) => {

		// I have to trigger my change off event.originalEvent in order to do stuff with data attributes in other places. 
		// Pretty annoying I can't dynamically add or read data attributes from `InputMask`
		const e = event.originalEvent.nativeEvent; 
		const val = e.target.value.replace(/[^0-9]/g, ''); // force only numbers to send to handleCreditCardMask()

		// save ccType value to session.reg.ccType in Redux store. This triggers the rerender when ccTypeState changes.
		dispatch(setRegProperty({ prop: 'ccType', val: getCreditCardTypeByNumber(val) }));

		Utils.inputChange(e); // A utils function that stores the inputted ccNumber value into the store, triggers off some other functionalilty based on data attributes and some other stuff, but for our purposes this is basically a dispatch to session.reg.ccNumber
	}

const CreditCardNumber1632 = ({
	ccNumberRef={ccNumberRef}
	inputBlur={Utils.inputBlur}
	inputChange={handleInputChange}
	inputFocus={Utils.inputFocus}
	mask={mask}
}) => {
	return (
		<div className="form-group">
			<InputMask
				autoClear={false}
				className="credit-card-number__input form-control"
				data-qa="credit-card-number-input"
				id="ccNumber"
				mask={mask}
				name="ccNumber"
				onBlur={(event) => inputBlur(event)}
				onChange={(event) => inputChange(event)}
				onFocus={(event) => inputFocus(event)}
				pattern="[0-9]*"
				ref={ccNumberRef}
				type="tel"
			/>
		</div>
	);
}
```

**Comment by melloware** [comment]
Instead of trying to switch the masks dynamically you should use a Credit Card component like: https://www.npmjs.com/package/react-credit-cards-2

**Comment by wlannistamps** [comment, marked as answer]
I have four other components that have various different masks to switch out based on different types of  input. Zip code, phone number, Credit Card, CVV. Trying to implement or find specific components for all of those needs is kind of nuts when all I need is a mask that can swap out dynamically. It's all good, I found what I needed with `react-input-mask`
